### PR TITLE
Fix for multiline errors when writing to log files

### PR DESCRIPTION
**Pull request by jemrobinson**
## :white_check_mark: Checklist

- [x] You have given your pull request a meaningful title (_e.g._ `Enable foobar integration` rather than `515 foobar`).
- [x] You are targeting the appropriate branch. If you're not certain which one this is, it should be **`develop`**.
- [x] Your branch is up-to-date with the **target branch** (it probably was when you started, but it may have changed since then).

### :vertical_traffic_light: Depends on

n/a

### :arrow_heading_up: Summary

Fix how multiline errors are written to a log file

**Previous**
```
2024-06-20 15:07:29,837 - ERROR - Timezone 'Mars/OlympusMons' was not recognised.
'Mars/OlympusMons'
```

**At c98ce56**
```
2024-06-20 15:07:29,837 - ERROR - Timezone 'Mars/OlympusMons' was not recognised.
2024-06-20 15:07:29,837 - ERROR - 'Mars/OlympusMons'
```

**At d08520d**
```
2024-06-20 15:07:29,837 - ERROR - Timezone 'Mars/OlympusMons' was not recognised.
                                  'Mars/OlympusMons'
```

### :closed_umbrella: Related issues

n/a

### :microscope: Tests

Tested locally

### CODE REVIEW

**Review by JimMadge — COMMENTED** [review]
I'm not sure if either of these are right.

One is difficult to parse as some lines are log messages, some are continuations.
The other suggests there were two errors.

Maybe lines should be joined?

**Comment by jemrobinson** [comment]
What would your desired output look like?

**Comment by JimMadge** [comment]
> What would your desired output look like?

Hmm, I'm not too sure. I'm not really familiar with log format conventions but I'm assuming trying to put multiline messages in a log might be frowned upon in the first place 😅.

I would think you'd want to keep to one message (even if it is long/multiline) per line in the logfile though.
Thinking you would want to be able to parse it like in `journalctl` to filter times, severity, _etc._

**Comment by jemrobinson** [comment]
If you really want to emit structured logs, we should probably be giving JSON output...

**Review by JimMadge — APPROVED** [review]

**Comment by jemrobinson** [comment]
I'm wondering whether we ought to do something like this instead?

```
2024-06-20 15:07:29,837 - ERROR - Timezone 'Mars/OlympusMons' was not recognised.\n'Mars/OlympusMons'
```

possibly combined with replacing

```
msg = f"Something.\n{exc}"
raise Exception(msg) from exc
```

with

```
msg = f"Something"
raise Exception(msg) from exc
```

in our own code?

**Comment by JimMadge** [comment]
That's what _I_ would have done as a first solution. Then see if very long lines become a big problem.

I suppose if would mean if you want to look through many lines you need an extra tool or bit of processing.

With tracebacks, I don't know if we need to print the previous exception in the new exception's message.

**Comment by jemrobinson** [comment]
Agreed - the insertion of newlines was done to ensure we got informative errors while skipping a full traceback. Note that other libraries (notably pydantic) also produce multiline errors, but we don't need to be doing it ourselves. I can update this PR next week - unless you want to do it as part of fixing logging?

File: data_safe_haven/administration/users/entra_users.py [hunk 1]
```diff
@@ -99,7 +99,7 @@ def list(self) -> Sequence[ResearchUser]:
                 for user_details in user_list
             ]
         except DataSafeHavenError as exc:
-            msg = f"Unable list Entra ID users.\n{exc}"
+            msg = "Unable list Entra ID users."
```
Comment (jemrobinson) [inline data_safe_haven/administration/users/entra_users.py]:
```suggestion
            msg = "Unable to list Entra ID users."
```

File: data_safe_haven/administration/users/entra_users.py [hunk 2]
```diff
@@ -114,7 +114,7 @@ def register(self, sre_name: str, usernames: Sequence[str]) -> None:
             for username in usernames:
                 self.graph_api.add_user_to_group(username, group_name)
         except DataSafeHavenError as exc:
-            msg = f"Unable add users to group '{group_name}'.\n{exc}"
+            msg = f"Unable add users to group '{group_name}'."
```
Comment (jemrobinson) [inline data_safe_haven/administration/users/entra_users.py]:
```suggestion
            msg = f"Unable to add users to group '{group_name}'."
```

File: data_safe_haven/external/api/azure_api.py [hunk 3]
```diff
@@ -724,7 +722,7 @@ def remove_blob(
                 f"Removed file [green]{blob_name}[/] from blob storage.",
             )
         except AzureError as exc:
-            msg = f"Blob file [green]'{blob_name}'[/] could not be removed from [green]'{storage_account_name}'[/].\n{exc}"
+            msg = f"Blob file [green]'{blob_name}'[/] could not be removed from [green]'{storage_account_name}'[/]."
```
Comment (jemrobinson) [inline data_safe_haven/external/api/azure_api.py]:
```suggestion
            msg = f"Blob file '{blob_name}' could not be removed from '{storage_account_name}'."
```

as these message now bypass our logger we probably don't want to inject colour formatting.

File: data_safe_haven/external/api/azure_api.py [hunk 4]
```diff
@@ -768,7 +766,7 @@ def remove_dns_txt_record(
                 f"Ensured that DNS record [green]{record_name}[/] is removed from zone [green]{zone_name}[/].",
             )
         except AzureError as exc:
-            msg = f"Failed to remove DNS record [green]{record_name}[/] from zone [green]{zone_name}[/].\n{exc}"
+            msg = f"Failed to remove DNS record [green]{record_name}[/] from zone [green]{zone_name}[/]."
```
Comment (jemrobinson) [inline data_safe_haven/external/api/azure_api.py]:
```suggestion
            msg = f"Failed to remove DNS record {record_name} from zone {zone_name}."
```

File: data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py [hunk 5]
```diff
@@ -63,7 +63,7 @@ def create(self, props: dict[str, Any]) -> CreateResult:
                 storage_account_name=props["storage_account_name"],
             )
         except Exception as exc:
-            msg = f"Failed to set ACLs on storage account [green]{props['storage_account_name']}[/].\n{exc}"
+            msg = f"Failed to set ACLs on storage account [green]{props['storage_account_name']}[/]."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py]:
```suggestion
            msg = f"Failed to set ACLs on storage account '{props['storage_account_name']}'."
```

File: data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py [hunk 6]
```diff
@@ -83,7 +83,7 @@ def delete(self, id_: str, props: dict[str, Any]) -> None:
                 storage_account_name=props["storage_account_name"],
             )
         except Exception as exc:
-            msg = f"Failed to delete custom ACLs on storage account [green]{props['storage_account_name']}[/].\n{exc}"
+            msg = f"Failed to delete custom ACLs on storage account [green]{props['storage_account_name']}[/]."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/blob_container_acl.py]:
```suggestion
            msg = f"Failed to delete custom ACLs on storage account '{props['storage_account_name']}'."
```

File: data_safe_haven/infrastructure/components/dynamic/entra_application.py [hunk 7]
```diff
@@ -59,7 +59,7 @@ def refresh(props: dict[str, Any]) -> dict[str, Any]:
                 )
             return outs
         except Exception as exc:
-            msg = f"Failed to refresh application [green]{props['application_name']}[/] in Entra ID.\n{exc}"
+            msg = f"Failed to refresh application [green]{props['application_name']}[/] in Entra ID."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/entra_application.py]:
```suggestion
            msg = f"Failed to refresh application '{props['application_name']}' in Entra ID."
```

File: data_safe_haven/infrastructure/components/dynamic/entra_application.py [hunk 8]
```diff
@@ -110,7 +110,7 @@ def create(self, props: dict[str, Any]) -> CreateResult:
                 else ""
             )
         except Exception as exc:
-            msg = f"Failed to create application [green]{props['application_name']}[/] in Entra ID.\n{exc}"
+            msg = f"Failed to create application [green]{props['application_name']}[/] in Entra ID."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/entra_application.py]:
```suggestion
            msg = f"Failed to create application '{props['application_name']}' in Entra ID."
```

File: data_safe_haven/infrastructure/components/dynamic/entra_application.py [hunk 9]
```diff
@@ -125,7 +125,7 @@ def delete(self, id_: str, props: dict[str, Any]) -> None:
             graph_api = GraphApi(auth_token=props["auth_token"])
             graph_api.delete_application(props["application_name"])
         except Exception as exc:
-            msg = f"Failed to delete application [green]{props['application_name']}[/] from Entra ID.\n{exc}"
+            msg = f"Failed to delete application [green]{props['application_name']}[/] from Entra ID."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/entra_application.py]:
```suggestion
            msg = f"Failed to delete application '{props['application_name']}' from Entra ID."
```

File: data_safe_haven/infrastructure/components/dynamic/entra_application.py [hunk 10]
```diff
@@ -156,7 +156,7 @@ def update(
             updated = self.create(new_props)
             return UpdateResult(outs=updated.outs)
         except Exception as exc:
-            msg = f"Failed to update application [green]{new_props['application_name']}[/] in Entra ID.\n{exc}"
+            msg = f"Failed to update application [green]{new_props['application_name']}[/] in Entra ID."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/entra_application.py]:
```suggestion
            msg = f"Failed to update application '{new_props['application_name']}' in Entra ID."
```

File: data_safe_haven/infrastructure/components/dynamic/file_share_file.py [hunk 11]
```diff
@@ -94,7 +94,7 @@ def create(self, props: dict[str, Any]) -> CreateResult:
             outs["file_name"] = file_client.file_name
         except Exception as exc:
             file_name = file_client.file_name if file_client else ""
-            msg = f"Failed to upload data to [green]{file_name}[/] in [green]{props['share_name']}[/].\n{exc}"
+            msg = f"Failed to upload data to [green]{file_name}[/] in [green]{props['share_name']}[/]."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/file_share_file.py]:
```suggestion
            msg = f"Failed to upload data to '{file_name}' in [green]{props['share_name']}[/]."
```

File: data_safe_haven/infrastructure/components/dynamic/file_share_file.py [hunk 12]
```diff
@@ -118,7 +118,7 @@ def delete(self, id_: str, props: dict[str, Any]) -> None:
                 file_client.delete_file()
         except Exception as exc:
             file_name = file_client.file_name if file_client else ""
-            msg = f"Failed to delete file [green]{file_name}[/] in [green]{props['share_name']}[/].\n{exc}"
+            msg = f"Failed to delete file [green]{file_name}[/] in [green]{props['share_name']}[/]."
```
Comment (jemrobinson) [inline data_safe_haven/infrastructure/components/dynamic/file_share_file.py]:
```suggestion
            msg = f"Failed to delete file '{file_name}' in [green]{props['share_name']}[/]."
```

File: data_safe_haven/external/api/azure_api.py (hunk 3 above)
Comment (JimMadge) [reply data_safe_haven/external/api/azure_api.py]: I think there should still be logged.

The `__init__` method of `DataSafeHavenError` calls `logger.error` with the exception message.

Comment (jemrobinson) [reply data_safe_haven/external/api/azure_api.py]: OK - does this interfere with the traceback output though?

Comment (JimMadge) [reply data_safe_haven/external/api/azure_api.py]: I would hope not, as that is also a rich feature. I'll try making a minimal example.

Comment (jemrobinson) [reply data_safe_haven/external/api/azure_api.py]: Great. If it works let's leave things as they are, if not let's remove the highlighting from exception messages. In either case, please merge once you're happy :).

Comment (JimMadge) [reply data_safe_haven/external/api/azure_api.py]: OK, so it looks like the answer is, the markup is processed by the called to log (through the rich handler) but the messages in exceptions and tracebacks are raw strings with the markup syntax.

We could do something like

```
class DataSafeHavenError(Exception):
    def __init__(self, message: str | bytes):
        super().__init__(remove_markup(message))

        # Log exception message as an error
        logger = get_logger()
        logger.error(message)
```

Comment (jemrobinson) [reply data_safe_haven/external/api/azure_api.py]: I don't think it's a problem to not have coloured output in our exception messages - the traceback is already colourful enough! Shall we just apply the suggestions to strip coloured formatting from the messages?

**Review by JimMadge — APPROVED** [review]

File: data_safe_haven/external/api/azure_api.py [hunk 13]
```diff
@@ -452,7 +450,7 @@ def ensure_storage_blob_container(
             )
             return container
         except HttpResponseError as exc:
-            msg = f"Failed to create storage container [green]{container_name}.\n{exc}"
+            msg = f"Failed to create storage container [green]{container_name}."
```
Comment (jemrobinson) [inline data_safe_haven/external/api/azure_api.py]:
```suggestion
            msg = f"Failed to create storage container '{container_name}'."
```

File: data_safe_haven/external/api/graph_api.py [hunk 14]
```diff
@@ -379,7 +379,7 @@ def ensure_application_service_principal(
                     raise DataSafeHavenMicrosoftGraphError(msg)
             return application_sp
         except Exception as exc:
-            msg = f"Could not create service principal for application '[green]{application_name}[/]'.\n{exc}"
+            msg = f"Could not create service principal for application '[green]{application_name}[/]'."
```
Comment (jemrobinson) [inline data_safe_haven/external/api/graph_api.py]:
```suggestion
            msg = f"Could not create service principal for application '{application_name}'."
```

**Pull request merged**